src/tls/connection.c

```diff
@@ -344,30 +356,155 @@ buffer_read_from_tls (Buffer           *self,
 }
 
 /**
- * connection_init_ws: Connect to a cockpit-ws instance for a new Connection
+ * https_instance_has_peer_cert: Check if that instance is for a given GnuTLS DER client certificate
+ *
+ * Returns true if either this ws instance has no client certificate and der is
+ * %NULL or empty, or if both certificates are identical. Otherwise returns false.
  */
 static bool
-connection_init_ws (Connection *self)
+https_instance_has_peer_cert (HTTPSInstance *inst, const gnutls_datum_t *der)
 {
-  struct sockaddr_un sockaddr = { .sun_family = AF_UNIX };
-  const char *sockname;
+  if (!der)
+    return inst->peer_cert.size == 0;
+  if (inst->peer_cert.size != der->size)
+    return false;
+  return memcmp (inst->peer_cert.data, der->data, der->size) == 0;
+}
+
+/**
+ * https_instance_new: Create a new cockpit-ws instance for given peer certificate
+ */
+static HTTPSInstance*
+https_instance_new (const gnutls_datum_t *der)
+{
+  struct sockaddr_un factory_sockaddr = { .sun_family = AF_UNIX };
+  size_t len;
   int r;
+  int fd;
+  HTTPSInstance *inst;
+
+  r = snprintf (factory_sockaddr.sun_path, sizeof factory_sockaddr.sun_path,
+                "%s/https-factory.sock", parameters.wsinstance_sockdir);
+  assert (r < sizeof factory_sockaddr.sun_path);
```

Comment:
Are we ok with using assert here?

Reply:
A call to warn and returning NULL might be more robust since wsinstance_sockdir is external input, no?

Reply:
I see, makes sense!  (Also, we wont ever defined NDEBUG, I guess.)